client/src/main/kotlin/io/spine/examples/pingh/client/LoginFlow.kt

```diff
@@ -0,0 +1,275 @@
+     * in scope of the device login flow.
+     */
+    public val verificationUrl: MutableStateFlow<Url> =
+        MutableStateFlow(event.verificationUrl)
+
+    /**
+     * The minimum duration that must pass before user can make a new access token request.
+     */
+    public val interval: MutableStateFlow<Duration> =
+        MutableStateFlow(event.interval)
+
+    /**
+     * Whether we can ask for a new user token from the external API.
+     *
+     * The contract of the external API assumes some delay that must pass
+     * before a new token can be requested. Therefore, we should wait for this call
+     * to become available.
+     *
+     * @see [interval]
+     */
+    public val canAskForNewTokens: MutableStateFlow<Boolean> = MutableStateFlow(true)
+
+    /**
+     * The duration after which the [userCode] expires.
+     */
+    public val expiresIn: MutableStateFlow<Duration> =
+        MutableStateFlow(event.expiresIn)
+
+    /**
+     * Whether the user code is expired.
+     *
+     * @see [expiresIn]
+     */
+    public val isUserCodeExpired: MutableStateFlow<Boolean> = MutableStateFlow(false)
+
+    /**
+     * Job that marks a [userCode] as expired after the [time][expiresIn] has passed.
+     */
+    private lateinit var codeExpirationJob: Job
+
+    init {
+        watchForCodeExpiration()
+    }
+
+    /**
+     * Starts a job that will mark the [userCode] as expired when [time][expiresIn] passes.
+     */
+    private fun watchForCodeExpiration() {
+        isUserCodeExpired.value = false
+        codeExpirationJob = invoke(expiresIn.value) {
+            isUserCodeExpired.value = true
+        }
+    }
+
+    /**
+     * Checks whether the user has completed the login on GitHub and entered their user code.
+     *
+     * @param onSuccess called when the login is successfully verified.
+     * @param onFail called when login verification fails.
+     */
+    public fun confirm(
+        onSuccess: (event: UserLoggedIn) -> Unit = {},
+        onFail: (event: UserIsNotLoggedIntoGitHub) -> Unit = {}
+    ) {
+        val command = VerifyUserLoginToGitHub::class.withSession(session.value!!.id)
+        client.observeEither(
+            EventObserver(command.id, UserLoggedIn::class) { event ->
+                codeExpirationJob.cancel()
+                onSuccess(event)
+            },
+            EventObserver(command.id, UserIsNotLoggedIntoGitHub::class) { event ->
+                preventAskingForNewTokens()
+                onFail(event)
+            }
+        )
+        client.send(command)
+    }
+
+    /**
+     * Prevents requests for new tokens during the [interval].
+     */
+    private fun preventAskingForNewTokens() {
+        canAskForNewTokens.value = false
+        invoke(interval.value) {
+            canAskForNewTokens.value = true
+        }
+    }
+
+    /**
+     * Requests new `UserCode` and updates state of the verification flow.
```

Comment:
From whom? Updates how?

Potential readers are not in your head. This is a common flaw, let's try to document it again.